### PR TITLE
Use New MultiOperation for MapShed and Sub-basin

### PR DESCRIPTION
**Pull request by rajadain**
## Overview

Switches MapShed and Sub-basin to use the new MultiOperation endpoint of the geoprocessing service introduced in https://github.com/WikiWatershed/mmw-geoprocessing/pull/85.

Connects #2657 

### Demo

The result of `debugcelery` for MapShed on this branch:

```
[2018-03-22 13:33:11,673: INFO/MainProcess] Received task: apps.modeling.geoprocessing.multi[6c735c63-3e1c-4c5f-ae9c-0ede5a23d567]
[2018-03-22 13:33:32,585: INFO/MainProcess] Received task: apps.modeling.mapshed.tasks.convert_data[7b92c891-7096-4851-9d63-1fb2d95ea5c3]
[2018-03-22 13:33:32,599: INFO/MainProcess] Received task: apps.modeling.mapshed.tasks.collect_data[9536ec97-e0be-47c1-9270-c7b0761e64fc]
[2018-03-22 13:33:33,100: INFO/MainProcess] Received task: apps.core.tasks.save_job_result[c1e63a96-8a1d-4c01-8e97-cb10453a9300]
```

as opposed to `develop`:

```
[2018-03-22 13:34:44,476: INFO/MainProcess] Received task: apps.modeling.geoprocessing.run[18856b80-89a3-4d73-a62d-25a3701e7456]
[2018-03-22 13:34:44,492: INFO/MainProcess] Received task: apps.modeling.geoprocessing.run[c0f7c1e6-3ab3-4295-9f11-f5929ed6d193]
[2018-03-22 13:34:44,500: INFO/MainProcess] Received task: apps.modeling.geoprocessing.run[7a2d54cd-0894-4a00-9410-563a6e1dcc9e]
[2018-03-22 13:34:44,513: INFO/MainProcess] Received task: apps.modeling.geoprocessing.run[78f5cdb3-ef07-4a27-beda-3ca95e6830e2]
[2018-03-22 13:34:44,532: INFO/MainProcess] Received task: apps.modeling.geoprocessing.run[a5caf95c-96c1-4f36-b541-0f727dfa9f1b]
[2018-03-22 13:34:44,573: INFO/MainProcess] Received task: apps.modeling.geoprocessing.run[769b1756-a480-4881-8b5d-5116dc24872f]
[2018-03-22 13:34:44,595: INFO/MainProcess] Received task: apps.modeling.geoprocessing.run[403a96b3-1571-4da9-9d68-5e4a6a09ae8a]
[2018-03-22 13:34:44,615: INFO/MainProcess] Received task: apps.modeling.geoprocessing.run[c375c572-21a9-4c9b-8703-d97dbcc49995]
[2018-03-22 13:34:44,636: INFO/MainProcess] Received task: apps.modeling.geoprocessing.run[8007fa43-7a4c-4fd6-9971-71ab12846a93]
[2018-03-22 13:34:44,651: INFO/MainProcess] Received task: apps.modeling.geoprocessing.run[a452f16a-ff6f-4904-bb51-ea0a391cd6f1]
[2018-03-22 13:34:44,666: INFO/MainProcess] Received task: celery.chord_unlock[4061869b-edc9-4f89-a168-a9fc2abe7e59]  ETA:[2018-03-22 17:34:45.442711+00:00]
[2018-03-22 13:34:44,677: INFO/MainProcess] Received task: apps.modeling.mapshed.tasks.nlcd_soil[83b30081-e83a-41c3-bcb9-930ed21a09ae]
[2018-03-22 13:34:44,690: INFO/MainProcess] Received task: apps.modeling.mapshed.tasks.soiln[2b5736ff-f117-45a1-96c7-ef6586278032]
[2018-03-22 13:34:44,700: INFO/MainProcess] Received task: apps.modeling.mapshed.tasks.soilp[4c051f5a-6011-4740-adb2-2483cf07378e]
[2018-03-22 13:34:44,715: INFO/MainProcess] Received task: apps.modeling.mapshed.tasks.recess_coef[dd9c4202-02a7-4ac2-8375-0fdbcc51c4ec]
[2018-03-22 13:34:44,725: INFO/MainProcess] Received task: apps.modeling.mapshed.tasks.gwn[27e24669-d645-4c5d-afab-2c8ade6cf965]
[2018-03-22 13:34:44,731: INFO/MainProcess] Received task: apps.modeling.mapshed.tasks.avg_awc[44d9d256-fa72-4f5c-af4a-7f17548efee1]
[2018-03-22 13:34:44,741: INFO/MainProcess] Received task: apps.modeling.mapshed.tasks.nlcd_slope[875f3c88-804e-4442-8591-b686a414bb6b]
[2018-03-22 13:34:44,748: INFO/MainProcess] Received task: apps.modeling.mapshed.tasks.slope[9ef9cd12-a45a-4869-b796-7a47b61e1273]
[2018-03-22 13:34:44,756: INFO/MainProcess] Received task: apps.modeling.mapshed.tasks.nlcd_kfactor[a22e0ef3-7af0-44cf-ba95-4b0a4e067175]
[2018-03-22 13:34:44,769: INFO/MainProcess] Received task: apps.modeling.mapshed.tasks.nlcd_streams[6030f3fa-3c34-42ea-8fe0-480654296ad7]
[2018-03-22 13:34:46,312: INFO/MainProcess] Received task: apps.modeling.mapshed.tasks.collect_data[e69c7324-b760-43fc-8a00-1c6763a51e7f]
[2018-03-22 13:34:46,835: INFO/MainProcess] Received task: apps.core.tasks.save_job_result[35a601b0-926e-4a2a-8f7d-f93bddda961d]
```

### Notes

We should make a follow up card to remove all the methods (such as `geoprocessing_chains`) that are not used anymore. All the individual callbacks (such as `slope`, `nlcd_kfactor`, etc) no longer need to be tasks and can have their `shared_task` decorators removed.

For sub-basin, we could have gone with the following workflow:

```
  chain(
    task0
    group(
      chain(task1, task2),
      chain(task3, task4))
    task5
    task6
  )
```

but because of celery/celery#4244 which does not give the first task in a chain in a group (task1 and task3 above) the result of the preceding task (task0 above), we couldn't have a setup like this:

```
  chain(
    multi_subbasin(huc12s)
    group(
      [convert_data(wkaoi) | collect_data(aoi, watershed_id)
       for (wkaoi, watershed_id, aoi) in huc12s])
    subbasin_results_to_dict
    save_job_result
  )
```

Instead we use `collect_subbasin` in place of the group for the same thing. If this part can be parallelized, it could lead to significant performance gains.

## Testing Instructions

* Check out this branch and reprovision `worker`
* Run `debugcelery`
* Make a MapShed project and observe the Celery logs. Ensure you see one `multi` instead of several `run` tasks. Ensure the values are the same for the same shape on `develop` / staging.
* Make a Sub-basin project and observe the Celery logs. Ensure you see one `multi` instead of several `run` tasks. Ensure the values are the same for the same shape on `develop` / staging.
* Try clearing the cache and running the above again to ensure it works without cached shapes:

    ```
    $ vagrant ssh services -c 'redis-cli -n 1 --raw KEYS ":1:geop_*" | xargs redis-cli -n 1 DEL'
    ```

* Try running the above again to ensure it works with cached shapes.

### CODE REVIEW

**Comment by rajadain** [comment]
@azavea-bot rebuild

**Comment by arottersman** [comment]
(1) I have commit I need to push that fixes a bug the current branch mapshedding drawn shapes

(2) As Terence notes, this is not the performance we expected given our new multi-op geoprocessing endpoint. The rabbit HUC-8 takes over a minute to complete just its mapshed portion, and will time out our UI. This could be especially worrisome because the rabbit has 62 or so huc-12s, and there huc-8s with close to a hundred.

The additional overhead comes from db calls. Geoprocessing now completes in 30sec, but the db calls to gather weather data, stream lengths, animals, etc 30+ secs themselves. Here's a depressing graph of what turned out to be relatively fast run:
screen shot 2018-03-27 at 6 56 24 pm

With the current set up we synchronously make 6 queries to populate the GMS for each huc-12. I've found the [weather data query](https://github.com/WikiWatershed/model-my-watershed/blob/586efe5c8454af386a5ba151b7ba876598d2bebd/src/mmw/apps/modeling/mapshed/calcs.py#L392), which takes a set of weather stations and calculates their average temps and precipitation values, to take disproportionately long (00.35sec for each huc12, so about 20sec altogether.) 

Here are what I believe are our paths forward:
1. Increase the frontend timeout
1. Decrease the allowed HUC size or limit sub-basin mode based on the number of contained huc-12s
1. Investigate re-implementing the queries such that they take a list of shapes and return a list of results. This would potentially cut down on IO costs. We could do this for all or some of the queries. The weather data related ones seem especially tricky given we get them via a two step process: 
    ```py
    # not real code
    # get top M weather stations near huc 12
    ws = nearest_weather_stations(huc12)
    # get precip and temp data for each weather station and calc averages
    wd = weather_data(ws)
    ```
     would need to become
    ```python
    list_of_ws = nearest_weather_stations(all_huc12s)
    list_of_wd = weather_data(list_of_ws)
    ```
1. With Barry's permission, investigate decreasing the weather station number in the case of sub-basin

File: src/mmw/apps/modeling/mapshed/calcs.py [hunk 1]
```diff
@@ -431,28 +437,39 @@ def f_to_c(f):
...
-            year = int(row[0]) - begyear
-            month = int(row[1]) - 1
+            station = int(row[0])
+            year = int(row[1]) - begyear
+            month = int(row[2]) - 1
             for day in range(31):
-                prcps[year][month][day] = float(row[day + 2]) * CM_PER_INCH
-
+                prcp = float(row[day + 3]) * CM_PER_INCH
+                prcps[station][year][month][day] = prcp
```
Comment (arottersman) [inline src/mmw/apps/modeling/mapshed/calcs.py]: This is resulting in different precipitation results than before. Temp is the same, but I don't think we've ever averaged the precipitation. It's throwing the whole model off.

Comment (mmcfarland) [reply src/mmw/apps/modeling/mapshed/calcs.py]: This surfaced a little while ago, last note is that @ajrobbins is going to follow up to determine if it's appropriate to average or not. See https://github.com/WikiWatershed/model-my-watershed/issues/2561 for previous discussion.

Comment (arottersman) [reply src/mmw/apps/modeling/mapshed/calcs.py]: That's great!! I'll leave it then.

**Comment by arottersman** [comment]
The changes on this PR are:
1) fixing mapshed for drawn shapes
2) adding statsd timers to the mapshed db queries, and the new geoprocessing operation
3) updating the weather query so that it happens once per subbasin run, and moving the averaging out to `numpy`. Some things about the changes are not ideal but get the job done with minimal changes, like `nearest_weather_data` now taking a list of undocumented "shape" tuples.

Here you can see the weather data changes halving the execution time for `collect_subbasin`:
screen shot 2018-03-29 at 1 02 53 pm

You can also see that the total for this HUC-8 is still a little over a minute. Before we work on further improvements we should review things on staging, and figure out what we'd have to increase the timeout to make it work (and if that's acceptable)


### Testing
* Test out a drawn AOI, and a well known AOI in regular mapshed. Confirm they complete.
* Compare a well known AOI (ie your favorite huc-12)'s mapshed results to staging. They should be the same except for the `Prec` dependent values. `Temp` key should be the same
* Run subbasin jobs on some huc-10s and confirm they complete
* Go to http://localhost:8080 and view the new graphs for the statsite timers 

screen shot 2018-03-30 at 9 27 53 am
screen shot 2018-03-30 at 9 27 46 am

File: src/mmw/apps/modeling/mapshed/calcs.py [hunk 2]
```diff
@@ -431,28 +437,39 @@ def f_to_c(f):
 
     year_range = endyear - begyear + 1
     stations = tuple([w.station for w in ws])
-    temps = [[[0] * 31 for m in range(12)] for y in range(year_range)]
-    prcps = [[[0] * 31 for m in range(12)] for y in range(year_range)]
+    temps = {s: [[[0] * 31 for m in range(12)]
```
[the remark this answers is not in the document]
Comment (arottersman) [reply src/mmw/apps/modeling/mapshed/calcs.py]: Thanks, I'll do both before merging!

**Pull request merged**